pkgs/jni/src/dartjni.c

```diff
@@ -464,9 +466,21 @@ Java_com_github_dart_1lang_jni_PortProxyBuilder__1invoke(
       destroy_cond(&result->cond);
     }
   } else {
+    // Flutter-specific: `mayEnterIsolate` is `true` when the proxy was
+    // constructed on the main thread and is being invoked on the main thread.
+    //
+    // When the current isolate is `null`, enter the main isolate that is pinned
+    // to the main thread first before invoking the `functionPtr`.
+    bool mustEnterIsolate = Dart_CurrentIsolate_DL() == NULL && mayEnterIsolate;
```

Comment:
LGTM. Though I just noticed another possible edge case.

If we're on the owner thread (`mayEnterIsolate` is true), and not entered into the target isolate, but instead entered into a different isolate. Then we should exit that other isolate, enter the target isolate, run the callback, exit the target isolate, then enter the other isolate. Of course, there's always the possibility that the other isolate is not owned by this thread, so could be entered by *another* thread while we're in the target isolate. What a mess.

Anyway, none of that is possible ATM, but could come up in future. Maybe just add an assert that `Dart_CurrentIsolate_DL()` is either null or `isolateId`?

Reply:
Done